Reintroduce mixing hard cap for all but the largest denom

UdjinM6 opened the pull request:
With many txes received on different addresses "nPrivateSendDenoms is not a hard cap but a checkpoint" logic introduced in #3479 creates hundreds of tiny denoms which are not easy to spend safely and user has no control over this behaviour anymore. This PR reverts it back to the original "nPrivateSendDenoms is a hard cap for most denoms". If someone wants "no hard cap for all denoms" in his wallet he should just bump `-privatesenddenoms` instead.

UdjinM6 commented:
If a user wants all his funds mixed he just have to bump `-privatesenddenoms`. Ignoring the hard cap basically brings back the behaviour which existed before #3434 by default but now even advanced users can't do anything about it because `-privatesenddenoms` has no effect and wallet just keeps creating useless dust. Example: receiving 8 txes 0.5 dash each with default settings (privatesenddenoms=50) results in 33x0.1, 63x0.01, 69x0.001. Note, that `-privatesenddenoms` was not reached for 0.1s yet exceeding 0.01s were created. So either the this statement https://github.com/dashpay/dash/blob/develop/src/privatesend/privatesend-client.cpp#L1597 is wrong or the algo is broken.

PastaPastaPasta commented:
The comment https://github.com/dashpay/dash/blob/develop/src/privatesend/privatesend-client.cpp#L1597 should likely read something "Now that nPrivateSendDenoms worth of each denom have been created or the max number of denoms given the value of the input, do something with the remainder."

The important aspect with the new algorithm is it's two stage nature, once nPrivateSendDenoms are created for a specific denom, it will create the bare minimum number of denoms of that denom, while still ensuring all funds are mixed.

The behavior is not the same as before or after #3434, if you ran your experiment before #3434 you would have a lot more 0.001 (probably, definitely would have a lot more if you received 20 .5 dash txes), so you'd have a LOT of tiny inputs but everything would get mixed. After #3434, you would only have 50 denoms, BUT you would have funds just sitting in your wallet not being mixed. 

However under the new algo, you have the best of both worlds, not a LOT of tiny inputs (algo creates as few as possible once wallet has 50 of that denom), but also all your funds will get mixed.

I want you to take a look at the individual create denomination txes in your wallet, you should see in the first tx or two, equal number of 0.001 and 0.01 denoms being created with a few 0.1 denoms, but in your last create denom tx there should be very few (less than 10) 0.001 denoms being created, but more 0.1 denoms being created

UdjinM6 commented:
What if I don't want all my coins to be split into a vast number of tiny denoms and would rather prefer them kept unmixed? If I don't need 100s of 0.01s and 0.001s and would rather keep just 20-ish of them just to be able to pay non-round amounts sometimes but most of my txes are much larger? How do I limit them now?

PastaPastaPasta commented:
I wouldn't call it "vast number of tiny denoms" it's the smallest amount possible while mixing all your funds.

I wouldn't be opposed to adding in a "hard cap" variable, at which point you'll have undenominated funds in your wallet, but I would want the to be by default effectively unlimited.

Also "hundreds" (say 200) of 0.001 denoms would require receiving at least 17 txes worst case txes ((200-50(nPrivateSendDenoms)) / 9 (worst case at saturation), and never spending your 0.001 denoms.

PastaPastaPasta commented:
@UdjinM6 you can see some testing here: https://docs.google.com/spreadsheets/d/1RsMJW_rJhHAIgntjFEu4AfBVc71Lv5mYWIq0JylnY94/edit?usp=sharing

Testing was all done with 10 inputs of 0.5 Dash each.

On develop:
```

End result
0.001 | 130 | (this is a worst case result over 10 txes, b/c an input of 0.5 creates the max number of 0.001 inputs at saturation, 9 per input
0.01 | 116 | (again, worst case, as 0.5 dash inputs creats 9 per input at saturation
0.1 | 37 | (theoretical max from 10 inputs of 0.5 dash would be 40)

Total amount which could be denominated but isn't | 0 Dash
0.00% Not mixed
```

Before #3434
```
End result

0.001 | 261
0.01 | 153
0.1 | 32

Total amount which could be denominated but isn't | 0 | Dash
0.00% | Not mixed
```

After #3434
```

End result

0.001 | 55
0.01 | 55
0.1 | 35

Total amount which could be denominated but isn't | 0.89288 | Dash in 10 inputs
17.86% | Not mixed
```

I didn't do the testing, but this PR seems to likely result in same performance as "After #3434" which in this test case would result in almost 20% not mixed

UdjinM6 commented:
Ok, so I dropped the original commit and reworked the whole thing based on what we discussed above. Also made few tweaks in the algo to respect limits a bit better. Pls see individual commits.

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1561,8 +1561,8 @@ bool CPrivateSendClientSession::CreateDenominated(CAmount nBalanceToDenominate,
 
-            // add each output up to 11 times or until it can't be added again or until we reach nPrivateSendDenoms
-            while (needMoreOutputs() && nOutputs <= 10 && currentDenomIt->second <= privateSendClient.nPrivateSendDenoms) {
+            // add each output up to 11 times or until it can't be added again or until we reach nPrivateSendDenomsBatched
```
I think "batched" should probably be called goal or something, batched doesn't really make sense to me

UdjinM6 replied:
agree, a4f5b66

PastaPastaPasta commented on src/wallet/init.cpp:
```diff
@@ -220,6 +221,19 @@ bool WalletInit::ParameterInteraction()
 
+    if (gArgs.IsArgSet("-privatesenddenoms")) {
+        int nDenomsDeprecated = gArgs.GetArg("-privatesenddenoms", DEFAULT_PRIVATESEND_DENOMS_HARDCAP);
+        InitWarning("Warning: -privatesenddenoms is deprecated, please use -privatesenddenomshardcap instead.\n");
```
should probably mention privatesenddenomsbatched here to. Something like "please instead use -privatesenddenomshardcap or -privatesenddenomsgoal depending on your purpose"

UdjinM6 replied:
80dff12

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1626,8 +1622,10 @@ bool CPrivateSendClientSession::CreateDenominated(CAmount nBalanceToDenominate,
                          "CPrivateSendClientSession::CreateDenominated -- 1 - totalOutputs: %d, nOutputsTotal: %d, nOutputs: %d, nValueLeft: %f\n",
                          nOutputsTotal + nOutputs, nOutputsTotal, nOutputs, (float) nValueLeft / COIN);
 
+                if (nOutputs + nOutputsTotal >= MAX_PRIVATESEND_DENOM_OUTPUTS) break;
             }
             nOutputsTotal += nOutputs;
+            if (nOutputsTotal >= MAX_PRIVATESEND_DENOM_OUTPUTS) break;
```
I don't see the point of adding these breaks, the 500 output check is primarily to ensure that the wallet doesn't become (overly) unresponsive, in actuality on develop it'll create txes up to 550 outputs, but I don't think that's a problem that needs to be fixed. Adding four breaks just doesn't seem worth it...

UdjinM6 replied:
agree partially, see 8b49159 kept them in "hard cap" cause there are no limits besides the balance otherwise

PastaPastaPasta commented on src/privatesend/privatesend-client.h:
```diff
@@ -35,6 +35,15 @@ static const int DEFAULT_PRIVATESEND_DENOMS_HARDCAP = 300;
+// More than 500 outputs starts to make qt quite laggy.
+// Additionally to need all 500 outputs (assuming a max per denom of 50) you'd need to be trying to
+// create denominations for over 3000 dash!
+static const int MAX_PRIVATESEND_DENOM_OUTPUTS = 500;
```
I'm not sure I'm a fan of extracting this... But I guess it's fine

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1581,7 +1581,7 @@ bool CPrivateSendClientSession::CreateDenominated(CAmount nBalanceToDenominate,
             // Check if this specific denom could use another loop, check that there aren't nPrivateSendDenomsBatched of this
             // denom and that our nValueLeft/nBalanceToDenominate is enough to create one of these denoms, if so, loop again.
-            if (it.second <= privateSendClient.nPrivateSendDenomsBatched && nValueLeft >= it.first && nBalanceToDenominate >= it.first) {
+            if (it.second < privateSendClient.nPrivateSendDenomsBatched && nValueLeft >= it.first && nBalanceToDenominate >= it.first) {
```
I like 75c17bfced90e50d0f1a9622290e08d9e0112043 :)

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1619,17 +1625,23 @@ bool CPrivateSendClientSession::CreateDenominated(CAmount nBalanceToDenominate,
+
+    for (const auto it : mapDenomCount) {
+        LogPrint(BCLog::PRIVATESEND,
+                "CPrivateSendClientSession::CreateDenominated -- 3 - DONE - nDenomValue: %f, count: %d\n",
+                (float) it.first / COIN, it.second);
```
same

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1578,13 +1578,19 @@ bool CPrivateSendClientSession::CreateDenominated(CAmount nBalanceToDenominate,
             // denom and that our nValueLeft/nBalanceToDenominate is enough to create one of these denoms, if so, loop again.
             if (it.second < privateSendClient.nPrivateSendDenomsBatched && nValueLeft >= it.first && nBalanceToDenominate >= it.first) {
                 finished = false;
+                LogPrint(BCLog::PRIVATESEND,
+                        "CPrivateSendClientSession::CreateDenominated -- 1 - NOT finished - nDenomValue: %f, count: %d, nValueLeft: %f, nBalanceToDenominate: %f\n",
+                        (float) it.first / COIN, it.second, (float) nValueLeft / COIN, (float) nBalanceToDenominate / COIN);
                 break;
             }
+            LogPrint(BCLog::PRIVATESEND,
+                    "CPrivateSendClientSession::CreateDenominated -- 1 - FINSHED - nDenomValue: %f, count: %d, nValueLeft: %f, nBalanceToDenominate: %f\n",
+                    (float) it.first / COIN, it.second, (float) nValueLeft / COIN, (float) nBalanceToDenominate / COIN);
```
I had this in my debugging code, but I feel like this will just spam debug.log for minimal value

UdjinM6 replied:
it's just few lines which do not repeat or spam that much really (comparing to everything else) but make it much easier to understand what's going on under the hood IMO

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1581,7 +1586,7 @@ bool CPrivateSendClientSession::CreateDenominated(CAmount nBalanceToDenominate,
             // Check if this specific denom could use another loop, check that there aren't nPrivateSendDenomsBatched of this
             // denom and that our nValueLeft/nBalanceToDenominate is enough to create one of these denoms, if so, loop again.
-            if (it.second < privateSendClient.nPrivateSendDenomsBatched && nValueLeft >= it.first && nBalanceToDenominate >= it.first) {
+            if (it.second < privateSendClient.nPrivateSendDenomsBatched && nValueLeft >= it.first && nBalanceToDenominate > 0) {
```
why? I added this check intentionally to check that the denom we are checking is small enough to actually create a denom based on balance left to denominate. With this change, I think this could get caught in an infinite loop

UdjinM6 replied:
pls see 374f2fd commit message, the "final" denom takes care of a potential infinite loop and fixes an excessive creation of an additional tx

PastaPastaPasta replied:
I never experienced this excessive creation in my testing.. Is there a specific thing you have to do?
